rejig the package installation instructions (#368)
springernature/frontend-playbook: practices/managing-node-projects.md

> # Managing Node.js-based projects and dependencies

Reviewer: Packages are listed as "dependencies" or "devDependencies" in the `package. json` and we start discussing the installation of these dependency packages straight away, so to me it makes more sense to use "dependencies". Also I think developers are more likely to have looked at a `package.json` than the npm docs.

     When running `npm install` to install `foo`'s dependencies, `npm` will warn about the problem.
     When `npm install`ing a package _which depends on_ `foo`, `npm` will warn and error.
     * Some deployment environments also respect it.
     * It makes compatibility requirements explicit to developers working on your application.
2. Using an `.nvmrc` file:
     * An `.nvmrc` file is a configuration file for `nvm` ([Node Version Manager](https://github.com/creationix/nvm)).
     Use of `nvm` is strongly encouraged as an application author, as it enables you to specify to other developers what versions of node they should be using when running your app.
     * Your application should include an [`.nvmrc` file](https://github.com/creationix/nvm#nvmrc) in the root directory of the project to specify which version(s) of node are compatible.
     * It also is required for proper `npm ci` usage, as explained above.
     * It also makes life much easier when working on multiple node projects locally that require different versions of node.

#### `nvm` usage

> When developers check out your project, they should first run `nvm use` before `npm ci` to ensure the project is running on the same version of node as you. Additionally [Travis respects `.nvmrc` files](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#specifying-nodejs-versions-using-nvmrc), so using one will simplify your Travis configuration.

Author: ```suggestion
When you check out a project, you should first run `nvm use` before `npm ci` to ensure you're running on the same version of node the project expects. Additionally [Travis respects `.nvmrc` files](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#specifying-nodejs-versions-using-nvmrc), so using one will simplify your Travis configuration and help other developers avoid mistakes.
```

#### Automatically running `nvm use`

As mentioned above, it is important to run `nvm use` before `npm ci`. But this is a thing that's easy to forget!

To avoid forgetting, there are useful shell extensions ([zsh](https://github.com/creationix/nvm#calling-nvm-use-automatically-in-a-directory-with-a-nvmrc-file),
[bash](https://stackoverflow.com/questions/23556330/run-nvm-use-automatically-every-time-theres-a-nvmrc-file-on-the-directory))
which run `nvm` and switch node versions for you when `cd`-ing into a directory. Well worth installing.

### Should I check in `package-lock.json` to version control?

Yes. There are two main reasons:

1. Predictable builds, as discussed above.

> 1. Dependency analysis tools (such as `npm audit`, GitHub Security Alerts and `snyk`) can spot insecure dependencies anywhere in the dependency tree by analysing the `package-lock.json` file. Furthermore these tools can force patch version updates of insecure dependencies way down the tree, rather than having to wait for all the package maintainers down that branch of the tree to release new packages with updated dependencies. This would not be possible without a `package-lock.json` and is a huge benefit.

Author: ```suggestion
1. Dependency analysis tools (such as `npm audit`, GitHub Security Alerts and `snyk`) can spot insecure dependencies anywhere in the dependency tree by analysing the `package-lock.json` file. These tools can force patch version updates of insecure dependencies in the tree, without waiting for third-party package maintainers to release new packages with updated dependencies. This wouldn't be possible without a `package-lock.json` and is a huge benefit.
```

The downside is it requires all developers to use `nvm` and `npm ci` to install dependencies, otherwise there will be constant conflicts in the `package-lock.json` file.

### Managing changes to the `package-lock.json` file

Firstly, bear in mind machine-generated files should not be hand-edited, and `package-lock.json` is no exception.

Secondly, if you are seeing unexpected merge conflicts in `package-lock.json` it can be a symptom of someone using `npm install` instead of `npm ci`. Speak to the developer and see if they are having trouble.

> Additionally, when updating your application its strongly recommend to keep `package-lock.json` changes as a result of `npm update` or `npm audit fix` to their own PR, free of other changes to the application. This helps by making your PR easier to review.

Author: ```suggestion
Additionally, when updating your application its strongly recommend to keep `package-lock.json` changes as a result of `npm update` or `npm audit fix` in a separate PR to other changes to the application. This helps by making your PRs easier to review.
```

Note that if a `package-lock.json` is present **`npm install` will not use it** as the source of authority for what to install.

This is the method we recommend for installing dependencies when authoring node _libraries_.

### `npm ci`

`npm ci` installs the complete dependency graph exactly as specified in the `package-lock.json` file, and so guarantees reproducible/consistent builds. This helps reduce "but it works on my machine" issues and is suitable for CI/CD environments (as the name implies).

[`npm ci` is also much quicker than `npm install`](https://docs.npmjs.com/cli/ci.html#description) if the `./node_modules` directory is not present (such as in a CI environment).

This is the method we recommend for installing dependencies when authoring node _applications_.

### How to specify the version of node to use

There are two main ways of doing this, and you're encouraged to do both.

1. The [`engines`](https://docs.npmjs.com/files/package.json#engines) field in `package.json`:

>      * `engines` is important if authoring libraries. Let's assume we're using a different version of node to that specified in `engines` field for package `foo`.

Author: ```suggestion
     * `engines` is important if you're authoring libraries. Let's assume we're using a different version of node to that specified in `engines` field for package `foo`.
```

>      When running `npm install` to install `foo`'s dependencies, `npm` will warn about the problem.

Author: ```suggestion
     When you run `npm install` to install `foo`'s dependencies, `npm` will warn about the problem.
```

     When `npm install`ing a package _which depends on_ `foo`, `npm` will warn.
     * Some deployment environments also respect it.
     * It makes compatibility requirements explicit to developers working on your application.
2. Using an `.nvmrc` file:
     * An `.nvmrc` file is a configuration file for `nvm` ([Node Version Manager](https://github.com/creationix/nvm)).

>      Use of `nvm` is strongly encouraged as an application author, as it enables you to specify to other developers what versions of node they should be using when running your app.

Author: ```suggestion
     We strongly encourage you to use `nvm` when you're authoring applications, as it enables you to specify to other developers what versions of node they should be using when running your app.
```

     * Your application should include an [`.nvmrc` file](https://github.com/creationix/nvm#nvmrc) in the root directory of the project to specify which version(s) of node are compatible.

>      * It also is required for proper `npm ci` usage, as explained above.

Reviewer 2: Should this be 'explained below'?

Reviewer: Nice spot thanks, fixed!

     * It also makes life much easier when working on multiple node projects locally that require different versions of node.

#### `nvm` usage

When you check out a project, you should first run `nvm use` before `npm ci` to ensure you're running on the same version of node the project expects. Additionally [Travis respects `.nvmrc` files](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#specifying-nodejs-versions-using-nvmrc), so using one will simplify your Travis configuration and help other developers avoid mistakes.

#### Automatically running `nvm use`

As mentioned above, it is important to run `nvm use` before `npm ci`. But this is a thing that's easy to forget!

> To avoid forgetting, there are useful shell extensions ([zsh](https://github.com/creationix/nvm#calling-nvm-use-automatically-in-a-directory-with-a-nvmrc-file),

Reviewer 3: would it be useful to suggest that a `preinstall` can be used to run `nvm use` before anything is installed? it wouldn't help with changing between projects, but it would help when installing modules for the first time and could be a useful defensive alternative that a developer could implement

Reviewer: Hmm, interesting point, thanks 🙂

I do get your argument, but in _this_ case I think it's probably better to crash early with errors on install TBH, to make developers aware of the issue. Anecdata but, I don't think I've ever used `preinstall` so this bit of magic might come as a bit of a surprise?

And yes as you say, it won't help with switching between projects (or the global node & npm versions being hiked). So am inclined to keep it simple...

Reviewer 3: Yep, that seems a fair point. Better to keep things simple

[bash](https://stackoverflow.com/questions/23556330/run-nvm-use-automatically-every-time-theres-a-nvmrc-file-on-the-directory))
which run `nvm` and switch node versions for you when `cd`-ing into a directory. Well worth installing.

### Should I check in `package-lock.json` to version control?

Yes. There are two main reasons:

1. Predictable builds, as discussed above.
1. Dependency analysis tools (such as `npm audit`, GitHub Security Alerts and `snyk`) can spot insecure dependencies anywhere in the dependency tree by analysing the `package-lock.json` file. These tools can force patch version updates of insecure dependencies in the tree, without waiting for third-party package maintainers to release new packages with updated dependencies. This wouldn't be possible without a `package-lock.json` and is a huge benefit.

The downside is it requires all developers to use `nvm` and `npm ci` to install dependencies, otherwise there will be constant conflicts in the `package-lock.json` file.

### Managing changes to the `package-lock.json` file

Firstly, bear in mind machine-generated files should not be hand-edited, and `package-lock.json` is no exception.

> Secondly, if you are seeing unexpected merge conflicts in `package-lock.json` it can be a symptom of someone using `npm install` instead of `npm ci`. Speak to the developer and see if they are having trouble.

Author: ```suggestion
Secondly, if you're seeing unexpected merge conflicts in `package-lock.json` it can be a symptom of someone using `npm install` instead of `npm ci`. Speak to the developer and see if they are having trouble.
```

> Additionally, when updating your application it's strongly recommended to keep `package-lock.json` changes as a result of `npm update` or `npm audit fix` in a separate PR to other changes to the application. This helps by making your PRs easier to review.

Author: ```suggestion
When you update your application, we strongly recommend that you keep `package-lock.json` changes as a result of `npm update` or `npm audit fix` in a separate PR to other changes to the application. This helps by making your PRs easier to review.
```

The following guide describes how we use node and manage package dependencies in [Springer Nature Node.js-based projects](https://github.com/springernature?utf8=%E2%9C%93&q=&type=public&language=javascript).

## Installing dependencies

> There are two methods for installing node dependencies: `npm install` and `npm ci`.

Reviewer: I disagree with your approach as there are several use cases;
 - "I just want to install the things" vs "i am seeking to actively develop this project"
 - apps vs libs
 - projects may or may not properly have a lockfile checked in

There are too many permutations to produce a prescriptive guide, and in any case I do not recall any principles of the runbook saying that we default to that approach.

"How to install node deps" seems a good place to start to teach these topics, rather than starting to talk about lockfiles, when lockfiles may or may not be relevant to my use case, and I may not even know what a lockfile is. For these reasons this doc is intended to teach concepts around `i` & `ci` and is more "educational" than most -- it is a complicated topic.

Note that if a `package-lock.json` is present **`npm install` will not use it** as the source of authority for what to install.

This is the method we recommend for installing dependencies when authoring node _libraries_.

### `npm ci`

`npm ci` installs the complete dependency graph exactly as specified in the `package-lock.json` file, and so guarantees reproducible/consistent builds. This helps reduce "but it works on my machine" issues and is suitable for CI/CD environments (as the name implies).

[`npm ci` is also much quicker than `npm install`](https://docs.npmjs.com/cli/ci.html#description) if the `./node_modules` directory is not present (such as in a CI environment).

This is the method we recommend for installing dependencies when authoring node _applications_.

### Using the right version of node

When authoring libraries you're probably supporting multiple versions of node, so you're not required to use a specific version when installing (as long as the version is supported by your library).

> When authoring node applications, and therefore using `npm ci`, it is important to specify a version of node to use. This is because the `package-lock.json` file will reflect the version of node used when the `package-lock.json` file was created.

Reviewer: I think it is true but am having a hard time finding anything beyond anecdata and random SO posts, so it may well be a superstition :)

I'm sure I remember problems with node-sass in this area, and fsevents, but I think these are examples where it _will or wont install_ according to _platform_. It's also true of course that packages may or may not install due to the user not using a supported version of node. But these are examples of a package either installing or not, and not examples of a successful install but of different _versions_ of the package.

Another thing to support that theory is that I cannot see a way to target different versions of packages at different versions of node.

So I think I should amend this line, and the following line, thanks!

> When authoring node applications, and therefore using `npm ci`, it is important to specify a version of node to use. This is because the `package-lock.json` file will reflect the version of node used when the `package-lock.json` file was created.

Reviewer: Have removed this section as I think its pointless.

If you don't use the same version of node when running `npm ci`, the versions of dependencies specified in the `package-lock.json` may be incompatible with the version of node you are currently using, and you may get installation errors.

### How to specify the version of node to use

There are two main ways of doing this, and you're encouraged to do both.

1. The [`engines`](https://docs.npmjs.com/files/package.json#engines) field in `package.json`:
     * `engines` is important if authoring libraries. Let's assume we're using a different version of node to that specified in `engines` field for package `foo`.
     When running `npm install` to install `foo`'s dependencies, `npm` will warn about the problem.

>      When `npm install`ing a package _which depends on_ `foo`, `npm` will warn and error.

Reviewer: Yep, will amend, thanks!

     * Some deployment environments also respect it.
     * It makes compatibility requirements explicit to developers working on your application.
2. Using an `.nvmrc` file:
     * An `.nvmrc` file is a configuration file for `nvm` ([Node Version Manager](https://github.com/creationix/nvm)).

>      Use of `nvm` is strongly encouraged as an application author, as it enables you to specify to other developers what versions of node they should be using when running your app.

Reviewer 4: > Use of `nvm` is strongly encouraged as an application author

We were talking about "applications" vs "libraries" before, the use of "application" here is confusing I think.

Reviewer: I am still talking about applications here.

     * Your application should include an [`.nvmrc` file](https://github.com/creationix/nvm#nvmrc) in the root directory of the project to specify which version(s) of node are compatible.
     * It also is required for proper `npm ci` usage, as explained above.
     * It also makes life much easier when working on multiple node projects locally that require different versions of node.

#### `nvm` usage

When you check out a project, you should first run `nvm use` before `npm ci` to ensure you're running on the same version of node the project expects. Additionally [Travis respects `.nvmrc` files](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#specifying-nodejs-versions-using-nvmrc), so using one will simplify your Travis configuration and help other developers avoid mistakes.

> #### Automatically running `nvm use`

Reviewer 4: I don't think this needs to be a separate point to "nvm usage"

Reviewer: Sorry, can you rephrase that please?

Reviewer: oh, i see what you mean. IIRC Hollie disagrees :) I think I do too. Using nvm is one topic, automatically using it is a different point.

### Should I check in `package-lock.json` to version control?

Yes. There are two main reasons:

1. Predictable builds, as discussed above.
1. Dependency analysis tools (such as `npm audit`, GitHub Security Alerts and `snyk`) can spot insecure dependencies anywhere in the dependency tree by analysing the `package-lock.json` file. These tools can force patch version updates of insecure dependencies in the tree, without waiting for third-party package maintainers to release new packages with updated dependencies. This wouldn't be possible without a `package-lock.json` and is a huge benefit.

The downside is it requires all developers to use `nvm` and `npm ci` to install dependencies, otherwise there will be constant conflicts in the `package-lock.json` file.

### Managing changes to the `package-lock.json` file

Firstly, bear in mind machine-generated files should not be hand-edited, and `package-lock.json` is no exception.

Secondly, if you are seeing unexpected merge conflicts in `package-lock.json` it can be a symptom of someone using `npm install` instead of `npm ci`. Speak to the developer and see if they are having trouble.

> Additionally, when updating your application its strongly recommend to keep `package-lock.json` changes as a result of `npm update` or `npm audit fix` in a separate PR to other changes to the application. This helps by making your PRs easier to review.

Reviewer 5: ```suggestion
Additionally, when updating your application it's strongly recommend to keep `package-lock.json` changes as a result of `npm update` or `npm audit fix` in a separate PR to other changes to the application. This helps by making your PRs easier to review.
```

  Your project should include an [`.nvmrc` file](https://github.com/creationix/nvm#nvmrc) in the root directory of the project to specify which version(s) of node are compatible.
  You can then run `nvm use` to use the right version of node. Additionally [Travis respects `.nvmrc` files](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#specifying-nodejs-versions-using-nvmrc), so using one will simplify your Travis configuration.

### Installing packages predictably

To reduce "but it works on my machine" issues, use a method that makes installs more predictable, and minimises changes to the `package-lock.json` (if your project is committing that file - this will be the case if your software is a node _application_, but not if it's a _library_).

#### Prefer `npm ci` to `npm install` where possible

> The `npm ci` command was introduced with `npm` version 5.7.0+ ([node v10.3+ or 8.12+](https://nodejs.org/en/download/releases/)). 

Reviewer 4: For simplicity, maybe we could remove this altogether. We shouldn't be using a version of Node.js that bundles a version of npm that doesn't support `npm ci`, so removing this would make the document less scary and easier to understand.

Reviewer: Agreed

If your project has a `package-lock.json` file (see [Should I check in `package-lock.json` to version control?](#should-i-check-in-package-lockjson-to-version-control)), **`npm ci` will always install predictably**, using the `package-lock.json` as a source of authority. [`npm ci` is also much quicker than `npm install`](https://docs.npmjs.com/cli/ci.html#description) if the `./node_modules` directory is not present (such as in a CI environment).

#### nvm users run `nvm use` before `npm install`

> If `npm ci` is unavailable (because you're using an older version of npm), or otherwise unsuitable (e.g. you have no `package-lock.json`, or you're installing new/updating existing dependencies), use `npm install`. 

Reviewer 6: ![image](https://user-images.githubusercontent.com/5796370/93756411-fe992080-fbfc-11ea-811c-695bc6d329c2.png)
 
This paragraph about falling back to `npm install` looks sandwiched inbetween the two paragraphs that should be next to each other that describe how to ensure the correct node version.

Perhaps move the paragraph down so it's just above the paragraph beginning with "Using npm install does't guarantee reproducible builds..."?

Author: oh yeah! good shout

> If `npm ci` is unavailable (because you're using an older version of npm), or otherwise unsuitable (e.g. you have no `package-lock.json`, or you're installing new/updating existing dependencies), use `npm install`. 

Reviewer 4: Same as above, I would change this so it suggests:

* `npm ci` for installing "prod ready" dependencies
* `npm install` to install new dependencies or bump transitive deps

Reviewer: Shall we get this PR merged and consider this later as well?
